include/swift/IDE/TypeCheckCompletionCallback.h

```diff
@@ -0,0 +1,84 @@
+//===--- TypeCheckCompletionCallback.h  -----------------------------------===//
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// Provides TypeCheckCompletionCallback implementations for the various kinds
+/// of code completion. These extract and persist information needed to compute
+/// completion results from the solutions formed during expression typechecking.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IDE_TYPECHECKCOMPLETIONCALLBACK_H
+#define SWIFT_IDE_TYPECHECKCOMPLETIONCALLBACK_H
+
+#include "swift/AST/Expr.h"
+#include "swift/AST/Type.h"
+#include "swift/Basic/LLVM.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace swift {
+class Decl;
+class DeclContext;
+class Type;
+class ValueDecl;
+class CodeCompletionExpr;
+
+namespace constraints {
+class ConstraintSystem;
+class Solution;
+} // namespace constraints
+
+namespace ide {
+
+class TypeCheckCompletionCallback {
+  bool GotCallback = false;
+
+public:
+  virtual ~TypeCheckCompletionCallback() {}
+
+  /// Called for each solution produced while type-checking an expression
+  /// that the code completion expression participates in.
+  virtual void sawSolution(const constraints::Solution &solution) {
+    GotCallback = true;
```

Comment:
This looks suspicious - there is no way to ensure that each override calls `TypeCheckCompletionCallback::sawSolution(solution)` internally. I see that they currently do but it only takes one to go unnoticed... I'd like to suggest `virtual` to be removed from this method, and instead be used on a `protected` `sawSolutionImpl(const Solution &)` which is going to be called from this one.

Reply:
Sounds like a good idea. I’ll do that.

Reply:
Implemented in https://github.com/apple/swift/pull/41885/commits/3597652960253685386103751dad1837fe410213